### PR TITLE
Add Dockerfile for running raiko-host and raiko-guest

### PR DESCRIPTION
**Pull request by pbeza**

### CODE REVIEW

**Comment by pbeza** [comment]
@johntaiko can you please review and check if that works for you? Just run:
```
docker-compose run --rm raiko
```
It starts `raiko-host` using port `9090` by default. You can send JSON requests to the running container with:
```
curl --location --request POST 'http://<CONTAINER_IP>:9090/' --header 'Content-Type: application/json' --data-raw '{                                         
  "jsonrpc": "2.0",                                                                                                                                                                                                                                                                                                                                                                   
  "id": 1,                                                                                                                                                                                
  "method": "proof",                                                                                                                                                                      
  "params": [                                                                                                                                                                             
    {                                                                                                                                                                                     
      "type": "Sgx",                                                                                                                                                                      
      "l2Rpc": "https://rpc.internal.taiko.xyz",                                             
      "l1Rpc": "https://l1rpc.internal.taiko.xyz",                                           
      "block": 2,                                                                                                                                                                         
      "prover": "0x70997970C51812dc3A010C7d01b50e0d17dc79C8",                                
      "graffiti": "0000000000000000000000000000000000000000000000000000000000000000"                                                                                                      
    }                                                                                                                                                                                                                                                                                                                                                                                 
  ]                                                                                                                                                                                       
}'
```

File: docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+# Docker layer for building Gramine
+# from https://github.com/gramineproject/gramine/blob/master/packaging/docker/Dockerfile
+#
+
+ARG UBUNTU_IMAGE=ubuntu:22.04
+
+FROM ${UBUNTU_IMAGE}
+
+ARG UBUNTU_CODENAME=jammy
+WORKDIR /usr/src/raiko
```
Comment (johntaiko) [inline docker/Dockerfile]: I think you can't place raiko in `/usr` https://tldp.org/LDP/Linux-Filesystem-Hierarchy/html/usr.html
`/opt/raiko` maybe be a good place https://tldp.org/LDP/Linux-Filesystem-Hierarchy/html/opt.html

File: docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+ENV IP_NUMBER=0.0.0.0
```
Comment (johntaiko) [inline docker/Dockerfile]:
```suggestion
ENV BIND=0.0.0.0:9090
```

File: docker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+ENV PORT_NUMBER=9090
+ENV SGX_INSTANCE_ID=123
+ENV LOG_PATH=/data/log/sgx
```
Comment (johntaiko) [inline docker/Dockerfile]: We can add a config file instead of environment for arguments

File: docker/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    ln -s ../../../raiko-guest/config/raiko-guest.manifest.template && \
+    gramine-manifest -Dlog_level=error -Darch_libdir=/lib/x86_64-linux-gnu/ raiko-guest.manifest.template raiko-guest.manifest && \
+    gramine-sgx-gen-private-key && \
+    gramine-sgx-sign --manifest raiko-guest.manifest --output raiko-guest.manifest.sgx
+
+CMD /restart_aesm.sh && \
+    cd raiko-host/guests/sgx && \
+    gramine-sgx ./raiko-guest bootstrap && \
+    cd - && \
+    RUST_LOG=debug cargo run --bin raiko-host -- --sgx-instance-id=${SGX_INSTANCE_ID} --bind=${IP_NUMBER}:${PORT_NUMBER} --log-path=${LOG_PATH}
```
Comment (johntaiko) [inline docker/Dockerfile]: `RUST_LOG=debug` Log level needs to be configurable, and we must move the final compiled executions into a new image environment without `rust` things.

File: docker/Dockerfile (hunk 3 above)
Comment (smtmfft) [reply docker/Dockerfile]: can use .env to share the vars between compose & dockerfile. I think at least the ports need export in compose.yml

Comment (johntaiko) [reply docker/Dockerfile]: We need to frequently modify configurations and the number of configuration items will continue to increase.
So, it's not a good idea to place our configurations in env.

File: docker/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    cp target/release/raiko-guest raiko-host/guests/sgx && \
+    cd raiko-host/guests/sgx && \
+    ln -s ../../../raiko-guest/config/raiko-guest.manifest.template && \
+    gramine-manifest -Dlog_level=error -Darch_libdir=/lib/x86_64-linux-gnu/ raiko-guest.manifest.template raiko-guest.manifest && \
+    gramine-sgx-gen-private-key && \
+    gramine-sgx-sign --manifest raiko-guest.manifest --output raiko-guest.manifest.sgx
+
+CMD /restart_aesm.sh && \
+    cd raiko-host/guests/sgx && \
+    gramine-sgx ./raiko-guest bootstrap && \
```
Comment (johntaiko) [inline docker/Dockerfile]: We can't run bootstrap on each run

Comment (johntaiko) [reply docker/Dockerfile]: Maybe we can let users pass the private key themselves.

File: raiko-guest/config/raiko-guest.manifest.template [hunk 6]
```diff
@@ -37,8 +37,8 @@ sgx.trusted_files = [
   "file:raiko-guest",
   "file:{{ gramine.runtimedir() }}/",
   "file:{{ arch_libdir }}/libgcc_s.so.1",
-  "file:{{ arch_libdir }}/libssl.so.3",
-  "file:{{ arch_libdir }}/libcrypto.so.3",
+  "file:{{ arch_libdir }}/libssl.so.1.1",
```
Comment (johntaiko) [inline raiko-guest/config/raiko-guest.manifest.template]: @pbeza I don't know if these versions need to be 3 exactly

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+# Build Stage
+FROM rust:1.75.0 as builder
+ARG BUILD_FLAGS=""
+WORKDIR /opt/raiko
+COPY . .
+RUN apt-get update && apt-get install -y cmake \
+    libclang-dev
+RUN cargo build --release ${BUILD_FLAGS}
+
+FROM gramineproject/gramine:1.6-focal as runtime
```
Comment (johntaiko) [inline Dockerfile]: Reuse the gramine image if we can split builder and runtime

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+COPY --from=builder /opt/raiko/target/release/raiko-host ./bin
+COPY ./sgx-ra/src/*.so /usr/lib/
+
+RUN cd ./guests/sgx && \
+    gramine-manifest -Dlog_level=error -Darch_libdir=/lib/x86_64-linux-gnu/ raiko-guest.manifest.template raiko-guest.manifest && \
+    gramine-sgx-gen-private-key && \
+    gramine-sgx-sign --manifest raiko-guest.manifest --output raiko-guest.manifest.sgx && \
+    cd -
+
+ENTRYPOINT [ "/opt/raiko/bin/raiko-host" ]
```
Comment (johntaiko) [inline Dockerfile]: Remove bootstrap, we don't want to bootstrap each start

File: docker/docker-compose.yml [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+        UBUNTU_CODENAME: jammy
+    image: raiko:latest
+    container_name: raiko
+    devices:
+      - "/dev/sgx_enclave:/dev/sgx_enclave"
+      - "/dev/sgx_provision:/dev/sgx_provision"
+    volumes:
+      - /data/log/sgx:/data/log/sgx
+      - /tmp/sgx:/tmp/sgx
+      - /secrets:/opt/raiko/secrets
```
Comment (johntaiko) [inline docker/docker-compose.yml]: Get bootstrap secret from user

**Comment by pbeza** [comment]
@johntaiko @smtmfft I think you can merge it now. Alternatively feel free to review it once again. Just please don't introduce any untested code changes anymore. Thanks!

To init/bootstrap SGX, run:
```
$ docker-compose run --rm raiko --init
```
Once bootstrapped, you can run `raiko-host` with:
```
$ docker-compose up raiko
```
Test with:
```
curl --location --request POST 'http://<CONTAINER_IP>:8085' --header 'Content-Type: application/json' --data-raw '{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "proof",
  "params": [
    {
      "type": "Sgx",
      "l2Rpc": "https://rpc.internal.taiko.xyz",
      "l1Rpc": "https://l1rpc.internal.taiko.xyz",
      "block": 2,
      "prover": "0x70997970C51812dc3A010C7d01b50e0d17dc79C8",
      "graffiti": "0000000000000000000000000000000000000000000000000000000000000000"
    }
  ]
}'
```
Just make sure to replace `CONTAINER_IP` with the IP of your container. You can check it with `docker ps` and [this magic command](https://stackoverflow.com/a/20686101/1321680):
```
docker inspect \
  --format '{{ .NetworkSettings.IPAddress }}' container_name_or_id
```
(there must be an easier way to interact with Docker container...)

**Pull request merged**